staging/src/k8s.io/api/resource/v1alpha3/types.go

```diff
@@ -0,0 +1,998 @@
+	// The spec is immutable.
+	Spec ResourceClaimSpec `json:"spec" protobuf:"bytes,2,name=spec"`
+
+	// Status describes whether the claim is ready for use.
+	// +optional
+	Status ResourceClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
+}
+
+// ResourceClaimSpec defines how a resource is to be allocated.
+type ResourceClaimSpec struct {
+	// Requests are individual requests for separate resources for the claim.
+	// An empty list is valid and means that the claim can always be allocated
+	// without needing anything. A class can be referenced to use the default
+	// requests from that class.
+	//
+	// +required
+	// +listType=atomic
+	Requests []Request `json:"requests" protobuf:"bytes,1,name=requests"`
+
+	// These constraints must be satisfied by the set of devices that get
+	// allocated for the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Constraints []Constraint `json:"constraints,omitempty" protobuf:"bytes,2,opt,name=constraints"`
+
+	// This field holds configuration for multiple potential drivers which
+	// could satisfy requests in this claim. It is ignored while allocating
+	// the claim.
+	//
+	// +optional
+	// +listType=atomic
+	Config []ClaimConfiguration `json:"config,omitempty" protobuf:"bytes,3,opt,name=config"`
+
+	// ControllerName defines the name of the DRA driver that is meant
+	// to handle allocation of this claim. If empty, allocation is handled
+	// by the scheduler while scheduling a pod.
+	//
+	// Must be a DNS subdomain and should end with a DNS domain owned by the
+	// vendor of the driver.
+	//
+	// This is an alpha field and requires enabling the DRAControlPlaneController
+	// feature gate.
+	//
+	// +optional
+	ControllerName *string `json:"controllerName,omitempty" protobuf:"bytes,4,opt,name=controllerName"`
+
+	// Future extension, ignored by older schedulers. This is fine because
+	// scoring allows users to define a preference, without making it a
+	// hard requirement.
+	//
+	// Score *SomeScoringStruct
+}
+
+// Request is a request for one of many resources required for a claim.
+// This is typically a request for a single resource like a device, but can
+// also ask for several identical devices. It might get extended to support
+// asking for one of several different alternatives.
+type Request struct {
+	// The name can be used to reference this request in a pod.spec.containers[].resources.claims
+	// entry and in a constraint of the claim.
+	//
+	// Must be a DNS label.
+	Name string `json:"name" protobuf:"bytes,1,name=name"`
+
+	*RequestDetail `json:",inline" protobuf:"bytes,2,name=requestDetail"`
+
+	// FUTURE EXTENSION:
+	//
+	// OneOf contains a list of requests, only one of which must be satisfied.
+	// Requests are listed in order of priority.
+	//
+	// +optional
+	// +listType=atomic
+	// OneOf []RequestDetail
+}
+
+// RequestDetail is embedded inside Request. Exactly one field must be set.
+type RequestDetail struct {
```

Comment:
Can a single claim request both devices and gizmos?

Reply:
Yes. But because they would be radically different, we wouldn't be able to do describe constraints that apply to both (https://github.com/kubernetes/kubernetes/pull/125488#discussion_r1659413405).

"partitionable devices" are not gizmos. They are still "devices" and have attributes, which is what is used for the constraint.

Reply:
Yes. I think I mentioned it also elsewhere: this is currently a bit odd because we have a one-of struct with just one entry. Conceptually in the one-of it's optional, but validation fails if it's not set, so it's also required. I'm not sure what the right way to express that is with `+optional` vs. `+required`.

I've changed it to `+optional`.